### DISCUSSION TITLE
Custom LifeSpanHandler for CefSharp.OffScreen

### DISCUSSION THREAD

**Discussion by L0lig1** [cefsharp/CefSharp#4909] [General]
Hello! I am using CefSharp.OffScreen to automate some tasks. 
On a page there is a button that opens a new tab with a PDF. 
I implemented ILifeSpanHandler, but when the button is clicked OnBeforePopup is not being triggered, as the breakpoint I have set is not hit. I am sure it is clicked as I create screenshots after each EvaluateJavaScriptAsync and the design of the button changes when clicked.
Have to say, that the page is very old and this is the link I am clicking: 
`
<a id="pdflink"
    role="menuitem" tabindex="-1" clickhandlerattached="clickHandlerAttached" href="#"
    javascript="javascript:__doPostBack('FormPage_view_LazyContainer$ctl01$ObjectFormPageDataSource_Locations_ListMenu','2')"></a>
`

I searched a lot, but did not find anything about this. Here's my code

CustomLifeSpanHandler:
```

public class CustomLifespanHandler : ILifeSpanHandler
{
    public bool DoClose(IWebBrowser chromiumWebBrowser, IBrowser browser)
    {
        return true;
    }

    public void OnAfterCreated(IWebBrowser chromiumWebBrowser, IBrowser browser)
    {
    }

    public void OnBeforeClose(IWebBrowser chromiumWebBrowser, IBrowser browser)
    {
    }

    bool ILifeSpanHandler.OnBeforePopup(IWebBrowser chromiumWebBrowser, IBrowser browser, IFrame frame, string targetUrl, string targetFrameName, WindowOpenDisposition targetDisposition, bool userGesture, IPopupFeatures popupFeatures, IWindowInfo windowInfo, IBrowserSettings browserSettings, ref bool noJavascriptAccess, out IWebBrowser newBrowser)
    {
        //browser.MainFrame.LoadUrl(targetUrl);
        newBrowser = null; // breakPoint here is not hitting
        return false;
    }
}
```

Program.cs

```
static void Main(string[] args)
{ 
    AsyncContext.Run(async delegate
    {
        Cef.EnableWaitForBrowsersToClose();
        var settings = new CefSettings();
        settings.CachePath = Path.GetFullPath("cache");
        var success = await Cef.InitializeAsync(settings);
        if (!success)
        {
            return;
        }
        using (var browser = new ChromiumWebBrowser("someurl"))
        {
            browser.LifeSpanHandler = new CustomLifespanHandler();
            await browser.WaitForInitialLoadAsync();
            // login
            // navigating to where the button is
            // clicking button here
            await browser.EvaluateScriptAsync($"document.getElementById('pdflink').click()");

            Cef.WaitForBrowsersToClose();
            Cef.Shutdown();
        }
    });
}
```

**Comment by amaitland** [comment]
You can use devtools to aid in debugging.

https://github.com/cefsharp/CefSharp/wiki/Trouble-Shooting#javascript-debugging

First confirm that your js is working as expected.

**Reply by L0lig1** [reply]
Hi! Sorry for the late answer, thanks for your quick response!
I am 100% sure that JS is working as expected. Here is a screenshot from the devtools. 
![grafik](https://github.com/user-attachments/assets/a6a1eec3-4833-481a-b360-5397a68b9ee3)
There are no errors being shown anywhere.
But as I mentioned, when it clicks the button that should open the popup, nothing happens. The breakpoint I have set is not hit
![grafik](https://github.com/user-attachments/assets/f6732a60-4726-4e12-9653-21397dd75997)

**Reply by amaitland** [reply]
> I am 100% sure that JS is working as expected. Here is a screenshot from the devtools.

Just because there wasn't an error doesn't anything. Did you try executing your javascript in the devtools console directly?

Try a stand alone example like https://www.w3schools.com/jsref/met_win_open.asp for testing purposes
